projects/output-and-error-handling/client/main.go

```
@@ -0,0 +1,92 @@
+	// This loop will run only if the number of reties doesn't exceed 3
+	for retries < 3 {
+		resp := getWeather()
+		// Close the response body after been fully read
+		defer resp.Body.Close()
+
+		// Read response body and store it in a var
+		body, err := io.ReadAll(resp.Body)
+		// Show error message if we cannot read the response body
+		if err != nil {
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprintf(os.Stderr, "Failed to read response body: %v\n", err)
+			os.Exit(1)
+		}
+
+		// Handle cases depending on the Status code of the response
+		switch resp.StatusCode {
+		case 200:
+			// Set the retries number to 0 so we can run the program many time when the response is 200
+			retries = 0
+
+			// Convert response body from binary to string
+			sb := string(body)
+			fmt.Fprint(os.Stdout, sb+"\n")
+			os.Exit(0)
+		case 429:
+			handleRateLimited(resp.Header.Get("Retry-After"), retries)
+		case 500:
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprint(os.Stderr, "Internal Server Error\n")
+			os.Exit(1)
+		default:
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprint(os.Stderr, "Unexpected Error\n")
+			os.Exit(1)
+		}
+	}
+}
+
+func getWeather() *http.Response {
+	// Connect to the server and getting a response
+	resp, err := http.Get("http://localhost:8080")
+	// Show error message if connection is not established
+	if err != nil {
+		fmt.Print("Sorry we cannot get the weather!\n")
+		fmt.Fprintf(os.Stderr, "Failed to make http request: %v\n", err)
+		os.Exit(1)
+	}
+	return resp
+}
+
+// Handle response and retry depending on the Retry-After header
+func handleRateLimited(retryTime string, retries int) {
+	retrySeconds := 0
+	var err error
+	retryTimeDate, err := time.Parse(time.RFC1123, retryTime)
+
+	if err == nil {
+		retrySeconds = int(time.Until(retryTimeDate).Seconds())
+	} else if retryTime == "a while" {
+		retrySeconds = 5
+	} else {
+		retrySeconds, err = strconv.Atoi(retryTime)
+		if err != nil {
```

You probably want to propagate an error here rather than handling it yourself